### PR TITLE
18 translation version name in title bar

### PR DESCRIPTION
**Pull request by b-w-cole**
## Description
A test to cover this test case: https://github.com/Sefaria/Sefaria-Playwright-Tests/issues/18

Includes:
* Parameterized tests for all Interface and Source language combinations
* Abstracted source language function for this and future purposes
* New Global variable for source language selection, pointing to DOM locators.

## Code Changes
_The following changes were made to the files below_

## Notes
Issue noted with Hebrew sources - the translation title doesn't appear by default. Such cases are commented out in the test parameters.

### CODE REVIEW

**Comment by b-w-cole** [comment]
@saengel , this one isn't quite ready to merge.  I'm wondering about your opinion on the file translation-version-name-appears-in-title.spec.ts, and the commented out tests.  Do you want me to test for the existence of a default translation title?  This will fail for Hebrew source language texts, as the default translation isn't selected in the translation side bar.

**Comment by saengel** [comment]
> @saengel , this one isn't quite ready to merge. I'm wondering about your opinion on the file translation-version-name-appears-in-title.spec.ts, and the commented out tests. Do you want me to test for the existence of a default translation title? This will fail for Hebrew source language texts, as the default translation isn't selected in the translation side bar.

Hi @b-w-cole - thank you again! Another fantastic PR - much appreciated!

I think for Hebrew source language texts, let's keep them commented out for now. Let me know when it's ready for a review.

**Comment by b-w-cole** [comment]
Hi @saengel ! Sorry for the long wait.  I tweaked the Hebrew source language tests to be their own separate test, as the logic needs to be slightly different.

However, I noticed a quirk with the Hebrew Interface / Hebrew Source translations.  

1. Load Hebrew Interface and Source text
2. Navigate to Translations
3. Select 'חומש רש״י, רבי שרגא זילברשטיין'
4. 'The Rashi chumash by Rabbi Shraga Silverstein' (in English) will appear in the translation title.

I would expect the Hebrew translation title to appear in the translation title.
image

Also the source text switches itself to Bilingual.
image

I will switch this test parameter to Shocken's translation for now and recommit.

**Comment by b-w-cole** [comment]
Hi @saengel , I was able to consolidate the English and Hebrew tests with a little rearranging.  Additionally, I added bilingual support.

File: e2e-tests/tests/translation-version-name-appears-in-title.spec.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+         // Change the Source Language of the text
+         await changeLanguageOfText(page, `${sourceLanguageToggle}`)
+ 
+         // Retain the translation name locator
+         const translationNameInTitle = page.locator('span.readerTextVersion')
+ 
+         // Navigate to the Translations sidebar by clicking on the text title
+         //Clicks on בראשית א׳ / Genesis I
+         await page.locator('h1').click()
```
Comment (saengel) [inline e2e-tests/tests/translation-version-name-appears-in-title.spec.ts:49]: Same comment throughout this file regarding locator use - I didn't comment in each location, just a general curiosity to see if it's possible to rewrite any of these.

File: e2e-tests/tests/translation-version-name-appears-in-title.spec.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+         
+         // Click on Translations
+         await page.getByRole('link', {name: `${translations}`}).click()
+         
+         // Wait for Translations side-bar to load by waiting for 'Translations' header
+         await page.waitForSelector('h3')
+
+        // Check if the default translation in the title matches the selected translation
+        // NOTE: We are skipping checking for the default translation here, due to the Hebrew text being default Masoretic
```
Comment (saengel) [inline e2e-tests/tests/translation-version-name-appears-in-title.spec.ts:58]: Can we rewrite to say "due to the Hebrew text not displaying a version title". That might be a more accurate description of what's going on in terms of site functionality (even if the root reason behind it is what you described here, for Tanakh texts at least)

File: e2e-tests/tests/translation-version-name-appears-in-title.spec.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        if(sourceLanguage !== 'Hebrew'){
+            const defaultTranslation = await translationNameInTitle.textContent()
+            await expect(page.locator('div.version-with-preview-title-line', {hasText: defaultTranslation!}).getByRole('link')).toHaveText(`${currentlySelected}`)
+        }
+        
+        // TODO: 4th translation, handling Hebrew Interface translations in Hebrew.  For example: 'חומש רש״י, רבי שרגא זילברשטיין' should appear in the translation title as written.
+        const translationNames = ['The Schocken Bible, Everett Fox, 1995 ©', '«Да» project']
+        
+        // Utilizing the traditional for-loop as there are async issues with foreach
```
Comment (saengel) [inline e2e-tests/tests/translation-version-name-appears-in-title.spec.ts:67]: Helpful comment! Thank you!

File: e2e-tests/utils.ts [hunk 4]
```diff
@@ -23,13 +23,16 @@ export const changeLanguage = async (page: Page, language: string) => {
 }
 
 export const goToPageWithLang = async (context: BrowserContext, url: string, language=DEFAULT_LANGUAGE) => {
-    if (!langCookies.length) {
-        const page: Page = await context.newPage();
-        await page.goto('');
-        await changeLanguage(page, language);
-        langCookies = await context.cookies();
-    }
+    // If a cookie already has contents, clear it so that the language cookie can be reset
```
Comment (saengel) [inline e2e-tests/utils.ts:26]: This is great, thank you again for the efforts to figure this out!

File: e2e-tests/tests/interface-language-is-sticky.spec.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    
+    // Hebrew Interface and Bilingual Source
+    {interfaceLanguage: 'Hebrew', interfaceLanguageToggle: LANGUAGES.HE, 
+        sourceLanguage: 'Bilingual', sourceLanguageToggle: SOURCE_LANGUAGES.BI, 
+        expectedSourceText: 'רֵאשִׁ֖ית בָּרָ֣א אֱלֹהִ֑ים אֵ֥ת הַשָּׁמַ֖יִם וְאֵ֥ת הָאָֽרֶץ׃', expectedBilingualText: 'When God began to create', expectedInterfaceText: 'מקורות' },
+    
+    // Hebrew Interface and Hebrew Source
+    {interfaceLanguage: 'Hebrew', interfaceLanguageToggle: LANGUAGES.HE, 
+        sourceLanguage: 'Hebrew', sourceLanguageToggle: SOURCE_LANGUAGES.HE, 
+        expectedSourceText: 'רֵאשִׁ֖ית בָּרָ֣א אֱלֹהִ֑ים אֵ֥ת הַשָּׁמַ֖יִם וְאֵ֥ת הָאָֽרֶץ׃', expectedBilingualText: '', expectedInterfaceText: 'מקורות' },
```
Comment (saengel) [inline e2e-tests/tests/interface-language-is-sticky.spec.ts:19]: Not a super huge deal (especially since copy and paste with Hebrew text into most IDEs can be really finicky) but the first letter of the verse (the ב) is missing in every instance of Hebrew `expectedSourceText`. Is it possible to add it in for clarity? Just so it makes more sense to people reading the tests.

File: e2e-tests/tests/interface-language-is-sticky.spec.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    {interfaceLanguage: 'English', sinterfaceLanguageToggle: LANGUAGES.EN, 
+        sourceLanguage: 'Bilingual', sourceLanguageToggle: SOURCE_LANGUAGES.BI, 
+        expectedSourceText: 'רֵאשִׁ֖ית בָּרָ֣א אֱלֹהִ֑ים אֵ֥ת הַשָּׁמַ֖יִם וְאֵ֥ת הָאָֽרֶץ׃', expectedBilingualText: 'When God began to create', expectedInterfaceText: 'Texts' },
+        
+    // English Interface and Hebrew Source
+    {interfaceLanguage: 'English', interfaceLanguageToggle: LANGUAGES.EN, 
+        sourceLanguage: 'Hebrew', sourceLanguageToggle: SOURCE_LANGUAGES.HE, 
+        expectedSourceText: 'רֵאשִׁ֖ית בָּרָ֣א אֱלֹהִ֑ים אֵ֥ת הַשָּׁמַ֖יִם וְאֵ֥ת הָאָֽרֶץ׃', expectedBilingualText: '', expectedInterfaceText: 'Texts' }
+
+].forEach(({interfaceLanguage, interfaceLanguageToggle, sourceLanguage, sourceLanguageToggle, expectedSourceText, expectedBilingualText, expectedInterfaceText}) => {
```
Comment (saengel) [inline e2e-tests/tests/interface-language-is-sticky.spec.ts:36]: Thank you for using `forEach` again, makes a big difference!

File: e2e-tests/tests/interface-language-is-sticky.spec.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        // Locating the source text segment, then verifying translation
+        await expect(page.locator('div.segmentNumber').first().locator('..').locator('p')).toContainText(`${expectedSourceText}`)
+
+        // Checking out the second part of the text, if 'Bilingual' is selected
+        if(`${sourceLanguage}` === 'Bilingual'){
+            await expect(page.locator('div.segmentNumber').first().locator('..').locator('p span').last()).toContainText(`${expectedBilingualText}`)
+        }
+    
+        // Validate Hebrew interface language is still toggled
+        const textLink = page.locator('a.textLink').first()
```
Comment (saengel) [inline e2e-tests/tests/interface-language-is-sticky.spec.ts:57]: Maybe consider renaming to `interfaceTextLink` for clarity?

File: e2e-tests/tests/translation-version-name-appears-in-title.spec.ts [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+         await page.waitForSelector('h3')
+
+        // Check if the default translation in the title matches the selected translation
+        // NOTE: We are skipping checking for the default translation here, due to the Hebrew text being default Masoretic
+        if(sourceLanguage !== 'Hebrew'){
+            const defaultTranslation = await translationNameInTitle.textContent()
+            await expect(page.locator('div.version-with-preview-title-line', {hasText: defaultTranslation!}).getByRole('link')).toHaveText(`${currentlySelected}`)
+        }
+        
+        // TODO: 4th translation, handling Hebrew Interface translations in Hebrew.  For example: 'חומש רש״י, רבי שרגא זילברשטיין' should appear in the translation title as written.
```
Comment (saengel) [inline e2e-tests/tests/translation-version-name-appears-in-title.spec.ts:64]: Do we need to leave this `TODO` here? Remind me what's outstanding (I thought we were skipping this case, due to the lack of version title being displayed)

**Review by saengel — CHANGES_REQUESTED** [review]
Hi @b-w-cole - overall looks really great! Some tricky cases you had to figure out here, and appreciate how you found a way... even when there were unexpected edge cases. 

I left a few comments, mostly for us to discuss (curious to hear your thoughts as well). We can go from there. As always, great work!

File: e2e-tests/tests/translation-version-name-appears-in-title.spec.ts (hunk 1 above)
Comment (b-w-cole) [reply e2e-tests/tests/translation-version-name-appears-in-title.spec.ts:49]: Hmm... I can change it to `await page.getByRole('heading').first().click()`. This still works, and it is using best practices.  What do you think?

File: e2e-tests/tests/translation-version-name-appears-in-title.spec.ts (hunk 2 above)
Comment (b-w-cole) [reply e2e-tests/tests/translation-version-name-appears-in-title.spec.ts:58]: Sure thing!  I will make this change.

File: e2e-tests/tests/interface-language-is-sticky.spec.ts (hunk 5 above)
Comment (b-w-cole) [reply e2e-tests/tests/interface-language-is-sticky.spec.ts]: Oh shoot, sorry about that!  It is fixed.

File: e2e-tests/tests/interface-language-is-sticky.spec.ts (hunk 7 above)
Comment (b-w-cole) [reply e2e-tests/tests/interface-language-is-sticky.spec.ts:65]: That makes sense to me.

File: e2e-tests/tests/interface-language-is-sticky.spec.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        const page = await goToPageWithLang(context,'/Genesis.1',`${interfaceLanguageToggle}`)
+        
+        // Clicking on the Source Language toggle
+        await page.getByAltText('Toggle Reader Menu Display Settings').click()
+
+        // Selecting Source Language
+        await page.getByRole('radiogroup', {name: 'Language'}).locator(`${sourceLanguageToggle}`).click()
+    
+        // Locating the source text segment, then verifying translation
+        await expect(page.locator('div.segmentNumber').first().locator('..').locator('p')).toContainText(`${expectedSourceText}`)
```
[the remark this answers is not in the document]
Comment (b-w-cole) [reply e2e-tests/tests/interface-language-is-sticky.spec.ts:53]: Ok I see.  What I was doing here was using the "Pick Locator" tool in VSCode.  Let me see if I can neaten it up.
image

[the remark this answers is not in the document]
Comment (b-w-cole) [reply e2e-tests/tests/interface-language-is-sticky.spec.ts:53]: I was able to condense this to `await expect(page.getByRole('region').locator('.segmentText span').last()).toContainText("expectedText")`, as well as remove some other locators, but I can only go so far, as I need to pinpoint exactly where the text is showing up.  There's not really roles associated with the divs and spans for text, so I use the CSS locators instead.  I could use getByText(), but if I do that, I cannot test.

Additionally, I was able to simplify the process for changing source language and remove those locators.

image

File: e2e-tests/tests/interface-language-is-sticky.spec.ts [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+        // Checking out the second part of the text, if 'Bilingual' is selected
+        if(`${sourceLanguage}` === 'Bilingual'){
+            await expect(page.locator('div.segmentNumber').first().locator('..').locator('p span').last()).toContainText(`${expectedBilingualText}`)
```
[the remark this answers is not in the document]
Comment (b-w-cole) [reply e2e-tests/tests/interface-language-is-sticky.spec.ts:57]: Thanks for keeping me on track with these.  I'm continuing to learn, and you're helping me out.  There's more I'd like to simplify in this for sure.  I'll let you know.

[the remark this answers is not in the document]
Comment (saengel) [reply e2e-tests/tests/interface-language-is-sticky.spec.ts:57]: Thank you!

File: e2e-tests/tests/translation-version-name-appears-in-title.spec.ts (hunk 1 above)
Comment (saengel) [reply e2e-tests/tests/translation-version-name-appears-in-title.spec.ts:49]: I think that would be great, thank you for the suggestion!

File: e2e-tests/tests/translation-version-name-appears-in-title.spec.ts (hunk 2 above)
Comment (saengel) [reply e2e-tests/tests/translation-version-name-appears-in-title.spec.ts:58]: Great, thank you!

File: e2e-tests/tests/interface-language-is-sticky.spec.ts (hunk 5 above)
Comment (saengel) [reply e2e-tests/tests/interface-language-is-sticky.spec.ts]: Perfect, no worries!

File: e2e-tests/tests/interface-language-is-sticky.spec.ts (hunk 7 above)
Comment (saengel) [reply e2e-tests/tests/interface-language-is-sticky.spec.ts:65]: Ok great!

File: e2e-tests/tests/interface-language-is-sticky.spec.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (b-w-cole) [reply e2e-tests/tests/interface-language-is-sticky.spec.ts]: Hello!  On this topic again, I took another look at this, and found that the best way is with the locator pointing to div classes.  I tried to use sensible CSS classes where it is clearer what the locator is pointing to. 

I am also submitting a correction to the commit I submitted a week ago.  I was thinking about it at the time that 

`await expect(page.getByRole('link', {name: currentlySelectedText}).locator('..')).toContainText(defaultTranslation!)`

would be appropriate for [translation-version-name-appears-in-title.spec.ts](https://github.com/Sefaria/Sefaria-Project/pull/1979/commits/e1d1f38493cedee56f20fa7a8f2de6b9edfbc239#diff-22850967213fc83d1f9326eccd4412e968bc952f69669f672805db46edef7fe5), but I'd like to change it.  While the test passes, it doesn't flow very well through the DOM.  Also, it is not at all recommended by the course I'm taking, so there it is that.  

image

Instead, I would like to change this line to be:

` await expect(page.locator('div.version-with-preview-title-line').filter({ hasText: currentlySelectedText })).toContainText(translationName!)`

It flows much more naturally in the DOM, looking through the document for all version title lines containing 'Currently Selected' and checking if this also contains the translation we want.  

image

What do you think?

[the remark this answers is not in the document]
Comment (saengel) [reply e2e-tests/tests/interface-language-is-sticky.spec.ts]: Hi Brandon, thank you for the thoughtful explanations and revisions here. I think this improvement is great. Again, we try to prefer user-facing attributes as much as possible - but sometimes we need to rely on locators and that should be fine. Thank you for these improvements.
